### PR TITLE
Sorting added batch queue variants

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve performance of `ORDER BY`, `MergeTree` merges, window functions using batch version of `BinaryHeap`.

### CODE REVIEW

**Comment by kitaisreal** [comment]
If there are only few values in columns. It is not final numbers this optimization also need to add optimization in `MergingSortedAlgorithm`.

```
SELECT DISTINCT Age FROM hits_100m_obfuscated;

┌─Age─┐
│   0 │
│  50 │
│  31 │
│  55 │
│  22 │
│  28 │
└─────
```
Before:
```
SELECT WatchID FROM hits_100m_obfuscated ORDER BY Age FORMAT Null;
0 rows in set. Elapsed: 4.154 sec. Processed 100.00 million rows, 900.00 MB (24.07 million rows/s., 216.64 MB/s.)
```
After:
```
SELECT WatchID FROM hits_100m_obfuscated ORDER BY Age FORMAT Null;
0 rows in set. Elapsed: 2.422 sec. Processed 100.00 million rows, 900.00 MB (41.28 million rows/s., 371.56 MB/s.)
```

**Comment by kitaisreal** [comment]
Numbers after added support into `MergingSortedAlgorithm`.
Before:
```
SELECT WatchID FROM hits_100m_obfuscated ORDER BY Age FORMAT Null;
0 rows in set. Elapsed: 4.154 sec. Processed 100.00 million rows, 900.00 MB (24.07 million rows/s., 216.64 MB/s.)
```
After:
```
SELECT WatchID FROM hits_100m_obfuscated ORDER BY Age FORMAT Null;
0 rows in set. Elapsed: 0.482 sec. Processed 100.00 million rows, 900.00 MB (207.47 million rows/s., 1.87 GB/s.)
```

File: src/Processors/Merges/Algorithms/MergingSortedAlgorithm.cpp [hunk 1]
```diff
@@ -87,133 +82,131 @@ void MergingSortedAlgorithm::consume(Input & input, size_t source_num)
...
 
-        /** And what if the block is totally less or equal than the rest for the current cursor?
-            * Or is there only one data source left in the queue? Then you can take the entire block on current cursor.
-            */
-        if (current.impl->isFirst()
-            && !current_inputs[current.impl->order].skip_last_row /// Ignore optimization if last row should be skipped.
-            && (queue.size() == 1
-                || (queue.size() >= 2 && current.totallyLessOrEquals(queue.nextChild()))))
+        bool limit_reached = false;
+        if (limit && merged_rows + updated_batch_size > limit)
```
Comment (novikd) [inline src/Processors/Merges/Algorithms/MergingSortedAlgorithm.cpp:137]: I guess it's just a little bit easier to read
```suggestion
        if (limit && ((merged_rows + updated_batch_size) > limit))
```

File: src/Core/SortCursor.h [hunk 2]
```diff
@@ -322,93 +330,154 @@ struct SortCursorWithCollation : SortCursorHelper<SortCursorWithCollation>
...
 
     bool isValid() const { return !queue.empty(); }
 
-    Cursor & current() { return queue.front(); }
+    Cursor & current() requires (strategy == SortingQueueStrategy::Default)
+    {
+        return queue.front();
+    }
+
+    std::pair<Cursor *, size_t> current() requires (strategy == SortingQueueStrategy::Batch)
```
Comment (novikd) [inline src/Core/SortCursor.h:376]: Why pointer, not a reference as for the default mode?

File: src/Core/SortCursor.h [hunk 3]
```diff
@@ -441,21 +514,65 @@ class SortingHeap
 
             child_it = begin + child_idx;
 
-            if ((child_idx + 1) < size && *child_it < *(child_it + 1))
+            if ((child_idx + 1) < size && (*child_it).greater(*(child_it + 1)))
```
Comment (novikd) [inline src/Core/SortCursor.h:517]: I'm not sure that's a problem, but before we had `operator <` (smaller) and now it's `greater`. Shouldn't you swap arguments?

**Review by novikd — APPROVED** [review]

File: src/Core/SortCursor.h (hunk 3 above)
Comment (kitaisreal) [reply src/Core/SortCursor.h:517]: It was confusing so I removed it. `operator <` called function `greater`.

File: src/Core/SortCursor.h (hunk 2 above)
Comment (kitaisreal) [reply src/Core/SortCursor.h:376]: Will try to change in follow up refactoring pull request. Not sure if pair can take reference as one of its values.

**Comment by kitaisreal** [comment]
@novikd thanks for review. Will fix style in follow up pull request also remove `operator <` from cursor.

**Pull request merged**